Allow using a button element for button blocks

aristath opened the pull request:
## What?
Allows using a `<button>` element for button blocks.

## Why?

Because it would improve semantics (and therefore also a11y), and allow us to use the block in more scenarios where a link is not the right element to use. 
Right now, the `core/button` blocks prints a link. Though visually this is OK, semantically it can be limiting and in some cases incorrect. Currently, we have no way to add an actual `<button>` element, and the button block doesn't do what someone with some basic HTML knowledge would expect it to do.
Additionally, a real `<button>` element is something we could use in other blocks like the search block, comments block, as well as other future blocks.

This came out of https://github.com/WordPress/gutenberg/pull/44214

## How?
Added a `tagName` attribute to the `core/button` block. `<button>` elements don't have `href`, so when using a button HTML element, the URL input is hidden from the UI.
UPDATE: After [this comment](https://github.com/WordPress/gutenberg/pull/54206#issuecomment-1708040803), A `type` attribute was also added to allow defining the button type. 

Nothing changes for link buttons.

IMPORTANT:
This PR does not add a UI for selecting the element. It just makes it possible for users to manually change it via the Code Editor, without getting an error when they switch back to the visual editor.

## Testing Instructions
1. Check any previous post that had a pre-existing button block, and make sure there are no errors in the console
2. In a post add a new button and check that it works as expected
3. Switch to the code editor, and add this:
```html
<!-- wp:buttons -->
<div class="wp-block-buttons"><!-- wp:button {"tagName":"button","type":"button"} -->
<div class="wp-block-button"><button type="button" class="wp-block-button__link wp-element-button">Actual button</button></div>
<!-- /wp:button --></div>
<!-- /wp:buttons --></div>
```
4. Switch back to the visual editor and confirm it works
5. Save the post, view it on the frontend of the site, inspect the buttons and confirm that they use the right element

### Testing Instructions for Keyboard
Nothing to test

aristath committed:
Allow using a button element for button blocks

aristath committed:
update tests / fixtures

fabiankaegy commented:
Question 🤔 

It is often recommended that any button element should have an explicit `type` set. Without it the element behaves differently depending on whether they are rendered inside of a `form` or not.

> The default value of type attribute for button HTML element is "submit" which is often not the desired behavior and may lead to unexpected page reloads. This rules enforces an explicit type attribute for all the button elements and checks that its value is valid per spec (i.e., is one of "button", "submit", and "reset").
 \- [eslint plugin react](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/button-has-type.md)

I can see how this automatic contextual awareness could also be seen as a feature rather than an issue. But then we run into the issue that we have no way of adding additional buttons inside forms like a reset button or just a regular button.

I would therefore say that we should add a `type` attribute as well here which would allow us to set the type alongside the `tagName`. 

```html
<!-- wp:buttons -->
<div class="wp-block-buttons"><!-- wp:button {"tagName":"button","type":"button"} -->
<div class="wp-block-button"><button class="wp-block-button__link wp-element-button" type="button">Actual button</button></div>
<!-- /wp:button --></div>
<!-- /wp:buttons --></div>
```

aristath committed:
Add "type" attribute for buttons

aristath commented:
Thank you for the feedback @fabiankaegy!

> I would therefore say that we should add a `type` attribute as well here which would allow us to set the type alongside the `tagName`.

Makes sense... Added, and updated the PR description 👍

fabiankaegy reviewed (approved):
The code looks good to me and in my testing this works as intended. I didn't run into any deprecation issues with it also.

Mamaduka commented on packages/block-library/src/button/block.json:
```diff
@@ -9,6 +9,14 @@
 	"keywords": [ "link" ],
 	"textdomain": "default",
 	"attributes": {
+		"tagName": {
+			"type": "string",
+			"default": "a"
+		},
```
This should probably be an `enum` as we only want to allow `a` and `button` elements.

aristath replied:
I don't think we can actually use `enum`... At least not without changing the [schema for block.json](https://github.com/WordPress/gutenberg/blob/7652de9ed159d4e9698e432c06b83e56777160c8/schemas/json/block.json#L106-L138) 🤔

Mamaduka replied:
What's the exact reason?

I thought we could use something like this:

https://github.com/WordPress/gutenberg/blob/6e9c1d3dc2e325b000494738dfeeaa66a407395e/packages/block-library/src/paragraph/block.json#L29-L32

aristath replied:
Ah OK!! That makes sense now... I thought you were talking about making the `type` an `enum` 😅

Mamaduka commented on packages/block-library/src/button/block.json:
```diff
@@ -9,6 +9,14 @@
 	"keywords": [ "link" ],
 	"textdomain": "default",
 	"attributes": {
+		"tagName": {
+			"type": "string",
+			"default": "a"
+		},
```
This should probably be an `enum` as we only want to allow `a` and `button` elements.

Mamaduka commented on packages/block-library/src/button/edit.js:
```diff
@@ -209,7 +220,7 @@ function ButtonEdit( props ) {
 					} }
 				/>
-				{ ! isURLSet && (
+				{ ! isURLSet && 'a' === TagName && (
```
Suggestion: The `'a' === TagName` check could be extracted into `isLinkTag` and reused in code.

Mamaduka commented on packages/block-library/src/button/save.js:
```diff
@@ -65,14 +69,15 @@ export default function save( { attributes, className } ) {
+				type={ 'button' === TagName ? buttonType : null }
 				className={ buttonClasses }
-				href={ url }
+				href={ 'button' === TagName ? null : url }
```
Same as above - `'button' === TagName` could be extracted into `isButtonTag` and reused.

Mamaduka reviewed (commented):
Sorry, I'm late to the party. This is a nice enhancement.

I left a few minor suggestions for a follow-up.

aristath commented:
@Mamaduka thank you for the review and notes!
A followup PR was created in #54419 with the tweaks mentioned above 👍 🎉
